vowpalwabbit/offset_tree.cc

```diff
@@ -0,0 +1,207 @@
+      }
+      while (tournaments.size() > 1)
+      {
+        // ids of next set of nodes start with the current
+        const auto num_tournaments = tournaments.size();
+        // node count (since ids are zero based)
+        auto id = nodes.size();
+        std::vector<uint32_t> new_tournaments;
+        new_tournaments.reserve(num_tournaments / 2 + 1);
+        for (size_t j = 0; j < num_tournaments / 2; ++j)
+        {
+          auto left = tournaments[2 * j];
+          auto right = tournaments[2 * j + 1];
+          new_tournaments.emplace_back(id);
+          nodes.emplace_back(id++, left, right, false);
+        }
+        if (num_tournaments % 2 == 1)
+          new_tournaments.emplace_back(tournaments.back());
+        tournaments = std::move(new_tournaments);
+      }
+      root_idx = tournaments[0];
+    }
+    catch (std::bad_alloc& e)
+    {
+      THROW("Unable to allocate memory for offset tree.  Label count:" << _num_leaf_nodes << " bad_alloc:" << e.what());
+    }
+    _initialized = true;
+  }
+
+  uint32_t min_depth_binary_tree::internal_node_count() const { return nodes.size() - _num_leaf_nodes; }
+
+  uint32_t min_depth_binary_tree::leaf_node_count() const { return _num_leaf_nodes; }
+
+  void offset_tree::init(uint32_t num_actions) { binary_tree.build_tree(num_actions); }
+
+  int32_t offset_tree::learner_count() const { return binary_tree.internal_node_count(); }
+
+  const std::vector<float>& offset_tree::predict(LEARNER::single_learner& base, example& ec)
+  {
+    // - pair<float,float> stores the scores for left and right nodes
+    // - prediction_buffer stores predictions for all the nodes in the tree for the duration
+    //   of the predict() call
+    // - static thread_local ensures only one copy per calling thread.  This is to reduce
+    //   memory allocations in steady state.
+    static thread_local std::vector<std::pair<float, float>> prediction_buffer(learner_count());
+    static thread_local std::vector<float> scores(binary_tree.leaf_node_count());
+
+    auto& t = binary_tree;
+
+    prediction_buffer.clear();
+    scores.resize(t.leaf_node_count());
+
+    // Handle degenerate cases of zero and one node trees
+    if (t.leaf_node_count() == 0)
+      return scores;
+    if (t.leaf_node_count() == 1)
+    {
+      scores[0] = 1.0f;
+      return scores;
+    }
+
+    // Get predictions for all internal nodes
+    for (auto idx = 0; idx < t.internal_node_count(); ++idx)
+    {
+      base.predict(ec, idx);
+      prediction_buffer.emplace_back(ec.pred.a_s[0].score, ec.pred.a_s[1].score);
+    }
+
+    // Compute action scores
+    const auto lrnr_strt_offset = t.leaf_node_count();
+    for (auto rit = t.nodes.rbegin(); rit != t.nodes.rend(); ++rit)
+    {
+      // done processing all internal nodes
+      if (rit->is_leaf)
+        break;
+
+      // update probabilities for left node
+      const float left_p = prediction_buffer[rit->id - lrnr_strt_offset].first;
```

Comment:
Why do we need to constantly use lrnr_strt_offset?

Reply:
min_depth_tree's node collection contains internal and leaf nodes.  lrnr_strt_offset is the begining of all the internal nodes.

Reply:
prediction_buffer only needs to hold the predictions for the internal nodes.

Reply:
Decided to hide the offset arithmetic behind a class.